### ISSUE TITLE
Kong 0.10.3 failed to use cassandra cluster

### ISSUE THREAD

**Issue by thochra** [Kong/kong#2757]
The problem seems localized on creation (POST) through the admin kong's API.
I can't tell if this can also happen for kong's frontend and/or other methods.

The problem can be reproduced for `apis`, `consumers` and `plugins` endpoints (at least).

We detect the problem when testing production containers (smartOS for cassandra + LX debian 8/jessie for Kong) on the servers.

### Summary
For all the tests we only used one Kong instance.

Tests done with 3 configurations:
 1- cassandra alone (=> no cluster mode activated)
 2- cassandra x3 Network topology (=> cluster replication strategy)
 3- cassandra x3 Simple strategy (=> cluster replication strategy)

Testing tools:
 1- kongfig (only 22 APIs creation, but also have consumers and plugins init)
 2- bash script capable of creating X apis endpoints in kong (only create APIs endpoints)

Note: same observation have been made with a bash script for consumers creation (consumer and jwt + acls configurations) which means there is a risk when creating consumers :(

## Use case cassandra alone: ##
Whatever the used tool no fails happened.
What we can observe is after the creation of the 200nth endpoint the creation slows down.
The bash were able to create 2000 APIs endpoint without any issue other than the creation slow down.

## Use case cassandra x3 Network topo (replication factor 2, dc1:2) ##
Kongfig:
We can observe error 500 coming back from kong randomly (means sometime it's on the first creation request, sometime later).

Bash script:
It acts like the kongfig tool.

Bash script with sleep of 1s between creations: Better but still fails. Prabably not related to high requests pressure.

## Use case cassandra x3 Simple strategy  (replication factor 2) ##
Kongfig:
same observation as for network topo strategy replication factor 2

Bash script:
same observation as for network topo strategy replication factor 2

Bash script with sleep of 1s between creations:
Endpoint #288 failed to create.
For one run the endpoint #1 also failed.
For one run the 300 endpoints were created correctly.

## Use case cassandra x3 Network topo (replication factor 1, dc1:1) ##
Kongfig:
Worked well.

Bash script:
Worked well.

## Use case cassandra x3 Simple strategy  (replication factor 1) ##
Kongfig:
Worked well.

Bash script:
Worked well.

## Use case cassandra x3 Network topo (replication factor 1, 2 DataCenters, dc1:1 and dc2:1) ##
Kongfig:
Failed with random 500 errors

Bash script:
Failed with random 500 errors

## Use case cassandra x3 Network topo (replication factor ~2, 2 DataCenters, dc1:2 and dc2:1) ##
Kongfig:
Failed with random 500 errors

Bash script:
Failed with random 500 errors.
Happens more often with this configuration.

## Use case cassandra x3 Network topo (replication factor ~2, 2 DataCenters, dc1:2,dc2:1, DCAwareRoundRobin => dc2) ##
Kong is configured with a dc aware round robin + is member of dc2 which hold only 1 Cassandra.

Kongfig:
Worked well.

Bash script:
Fail later, usually after more 100+ APIs endpoints creation.

## KONG 0.11.0rc2 Use case cassandra x3 Network topo (replication factor ~2, 2 DataCenters, dc1:2 and dc2:1) ##
Tested with Kong performing a round robin and DC aware round robin over cassandra instances.

Kongfig:
Worked well.

Bash script:
Worked well.

APIs usage test:
After patching the file `/usr/local/share/lua/5.1/kong/core/plugins_iterator.lua` with the 0.11.0rc1 version of the file, the APIs worked well.

Note: Migrations (Cassandra schema preparation) have to do be ran manually, but it works better than with 0.10.3.
      No migration failed during the multiple tries I did.

## Example of error obtained when receiving the 500 ##
For all cases (apis, plugins, consumers) it was related to a `nil` value into a `marshall_event`

```
2017/07/31 13:27:39 [error] 8964#0: *49591 lua coroutine: runtime error: /usr/local/share/lua/5.1/kong/dao/schemas/plugins.lua:55: attempt to index local 'plugin_t' (a nil value)
stack traceback:
coroutine 0:
        /usr/local/share/lua/5.1/kong/dao/schemas/plugins.lua: in function 'marshall_event'
        /usr/local/share/lua/5.1/kong/dao/dao.lua:68: in function 'event'
        /usr/local/share/lua/5.1/kong/dao/dao.lua:134: in function 'insert'
        /usr/local/share/lua/5.1/kong/api/crud_helpers.lua:167: in function 'post'
        /usr/local/share/lua/5.1/kong/api/routes/apis.lua:48: in function </usr/local/share/lua/5.1/kong/api/routes/apis.lua:47>
coroutine 1:
        [C]: in function 'resume'
        /usr/local/share/lua/5.1/lapis/application.lua:393: in function 'handler'
        /usr/local/share/lua/5.1/lapis/application.lua:130: in function 'resolve'
        /usr/local/share/lua/5.1/kong/api/init.lua:65: in function 'handler'
        /usr/local/share/lua/5.1/lapis/application.lua:130: in function 'handler'
        /usr/local/share/lua/5.1/lapis/application.lua:163: in function </usr/local/share/lua/5.1/lapis/application.lua:159>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/lapis/application.lua:159: in function 'dispatch'
        /usr/local/share/lua/5.1/lapis/nginx.lua:214: in function 'serve'
        content_by_lua(nginx-kong.conf:161):10: in function <content_by_lua(nginx-kong.conf:161):1>, client: 10.130.1.32, server: kong_admin, request: "POST /apis/3a8b608a-5a72-4813-b35f-c3dac7dc69a6/plugins HTTP/1.1", host: "kong-01.ms.virtua.ch:8001"
```

### Steps To Reproduce

1. Check that gist for information on what I played with: https://gist.github.com/thochra/fbfe8c42072a91fcf934d9b0cd737d74

### Additional Details & Logs

- Kong version 0.10.3
- Operating System: docker kong:0.10.3

**Comment by thochra** [comment]
For information moved on 0.11.0rc2 because the issue seems to not happening there.

**Comment by thibaultcha** [comment]
Closing seems as per your comment, this seems to be resolved in Kong 0.11. Thanks!
